src/Confluent.SchemaRegistry/CachedSchemaRegistryClient.cs

```diff
@@ -0,0 +1,215 @@
+            if (schemaRegistryUrisMaybe.Value == null)
+            {
+                throw new ArgumentException("schema.registry.url configuration property must be specified.");
+            }
+            var schemaRegistryUris = (string)schemaRegistryUrisMaybe.Value;
+
+            var timeoutMsMaybe = config.Where(prop => prop.Key.ToLower() == SchemaRegistryConnectionTimeoutMsPropertyName).FirstOrDefault();
+            var timeoutMs = timeoutMsMaybe.Value == null ? DefaultTimeout : (int)timeoutMsMaybe.Value;
+
+            var identityMapCapacityMaybe = config.Where(prop => prop.Key.ToLower() == SchemaRegistryMaxCachedSchemasPropertyName).FirstOrDefault();
+            this.identityMapCapacity = identityMapCapacityMaybe.Value == null ? DefaultMaxCapacity : (int)identityMapCapacityMaybe.Value;
+
+            foreach (var property in config)
+            {
+                if (property.Key != SchemaRegistryUrlPropertyName && 
+                    property.Key != SchemaRegistryConnectionTimeoutMsPropertyName && 
+                    property.Key != SchemaRegistryMaxCachedSchemasPropertyName)
+                {
+                    throw new ArgumentException($"CachedSchemaRegistryClient: unexpected configuration parameter {property.Key}");
+                }
+            }
+
+            this.restService = new RestService(schemaRegistryUris, timeoutMs);
+        }
+
+        /// <remarks>
+        ///     Ensure memory doesn't explode in the case of incorrect usage.
+        /// </remarks>
+        private void CheckIfCacheFull()
+        {
+            // TODO: an LRU cache is arguably better, though if the cache fills up this
+            // will usually be unexpected behavior, so failing fast is actually pretty 
+            // good.
+            if (this.schemaById.Count >= identityMapCapacity)
+            {
+                throw new OutOfMemoryException("Local schema cache maximum capacity exceeded");
+            }
+        }
+
+        /// <include file='include_docs.xml' path='API/Member[@name="ISchemaRegistryClient_GetSchemaIdAsync"]/*' />
+        public async Task<int> GetSchemaIdAsync(string subject, string schema)
+        {
+            if (!this.idBySchemaBySubject.TryGetValue(subject, out Dictionary<string, int> idBySchema))
+            {
+                idBySchema = new Dictionary<string, int>();
+                this.idBySchemaBySubject[subject] = idBySchema;
+            }
+
+            // TODO: This could be optimized in the usual case where idBySchema only
+            // contains very few elements and the schema string passed in is always
+            // the same instance.
+
+            if (!idBySchema.TryGetValue(schema, out int schemaId))
+            {
+                CheckIfCacheFull();
+
+                schemaId = (await restService.CheckSchemaAsync(subject, schema, true)).Id;
+                idBySchema[schema] = schemaId;
+                schemaById[schemaId] = schema;
```

Comment:
this stuff needs to be protected by a lock, right?

Reply:
yep, you're right.